numpy/core/src/multiarray/ctors.c

```diff
@@ -801,14 +805,29 @@ PyArray_NewFromDescr_int(
                 fa->strides[i] = 0;
             }
         }
-        /*
-         * It is bad to have uninitialized OBJECT pointers
-         * which could also be sub-fields of a VOID array
-         */
-        if (zeroed || PyDataType_FLAGCHK(descr, NPY_NEEDS_INIT)) {
+
+        if (
+            /*
+             * If NPY_NEEDS_INIT is set, always zero out the buffer, even if
+             * zeroed isn't set.
+             */
+            (PyDataType_FLAGCHK(descr, NPY_NEEDS_INIT)) ||
+            /*
+             * If fill_zero_value isn't set, the dtype definitely has no
+             * special zero value, so get a zero-filled buffer using
+             * calloc.  Otherwise, get the buffer with malloc and allow
+             * fill_zero_value to zero out the array with the appropriate
+             * special zero value for the dtype. VOID arrays have
+             * fill_zero_value set but may or may not contain objects, so
+             * always allocate with calloc if zeroed is set.
```

Comment:
If they contain objects, they should have `NPY_NEEDS_INIT` set?

Reply:
They should. The comment is referring to operations like `np.zeros(shape=10, dtype=[("A", "M8[s]")])`. It doesn't contain objects but it has `fill_zero_value` set since it's an `NPY_VOID` array, so unless I do something special for void dtypes the array buffer would get allocated with `malloc`. Is there a way I can rewrite the comment to make that clearer?

Reply:
Hmmmm... I think the important point is that `PyArray_FillObjectArray` which is called later assumes that everything not object is already zeroed?  So that is why you need to zero here for that case.

(That seems like somewhat weird logic, but lets look at it after the other changes, it might just go away by then anyway.)